File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/utils/ConsistencyUtils.java
package ru.kmorozov.library.data.loader.utils;

import org.springframework.data.repository.CrudRepository;
import ru.kmorozov.library.data.model.book.Book;

import java.util.*;

/**
 * Created by sbt-morozov-kv on 07.04.2017.
 */
public class ConsistencyUtils {

    private static final boolean DEDUPLCATION_ENABLED = false;

    public static Collection<Book> deduplicate(Collection<Book> books, CrudRepository booksRepository) {
        if (!ConsistencyUtils.DEDUPLCATION_ENABLED)
            return books;

        Collection<String> uniquePaths = new ArrayList<>();
        Collection<Book> deduplicatedBooks  = new ArrayList<>();
        for (Book book : books)
            if (uniquePaths.add(book.getBookInfo().getPath()))
                deduplicatedBooks.add(book);
            else
                booksRepository.delete(book);

        return deduplicatedBooks;
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/utils/ConsistencyUtils.kt
package ru.kmorozov.library.data.loader.utils

import org.springframework.data.repository.CrudRepository
import ru.kmorozov.library.data.model.book.Book

import java.util.*

/**
 * Created by sbt-morozov-kv on 07.04.2017.
 */
object ConsistencyUtils {

    private const val DEDUPLCATION_ENABLED = false

    fun deduplicate(books: Collection<Book>, booksRepository: CrudRepository<Book, String>): Collection<Book> {
        if (!DEDUPLCATION_ENABLED)
            return books

        val uniquePaths = ArrayList<String>()
        val deduplicatedBooks = ArrayList<Book>()
        for (book in books)
            if (uniquePaths.add(book.bookInfo.path!!))
                deduplicatedBooks.add(book)
            else
                booksRepository.delete(book)

        return deduplicatedBooks
    }
}
